src/main/java/org/veupathdb/service/eda/ds/plugin/standalone-map/StandaloneMapMarkersPlugin.java

```diff
@@ -0,0 +1,202 @@
+  @Override
+  protected void writeResults(OutputStream out, Map<String, InputStream> dataStreams) throws IOException {
+
+    // create scanner and line parser
+    InputStreamReader isReader = new InputStreamReader(new BufferedInputStream(dataStreams.get(DEFAULT_SINGLE_STREAM_NAME)));
+    BufferedReader reader = new BufferedReader(isReader);
+    DelimitedDataParser parser = new DelimitedDataParser(reader.readLine(), TAB, true);
+
+    // establish column header indexes
+    MapSpec spec = getPluginSpec();
+    Function<VariableSpec,Integer> indexOf = var -> parser.indexOfColumn(getUtil().toColNameOrEmpty(var)).orElseThrow();
+    int geoVarIndex = indexOf.apply(spec.getGeoAggregateVariable());
+    int latIndex    = indexOf.apply(spec.getLatitudeVariable());
+    int lonIndex    = indexOf.apply(spec.getLongitudeVariable());
+
+    ParsedGeolocationViewport viewport = ParsedGeolocationViewport.fromApiViewport(spec.getViewport());
+    long entityRecordsWithGeoVar = 0;
+
+    // loop through rows of data stream, aggregating stats into a map from aggregate value to stats object
+    Map<String, GeoVarData> aggregator = new HashMap<>();
+    String nextLine = reader.readLine();
+
+    while (nextLine != null) {
+      String[] row = parser.parseLineToArray(nextLine);
+      
+      // entity records counts not impacted by viewport
+      if (!(row[geoVarIndex] == null ||
+            row[geoVarIndex].isEmpty() ||
+            row[latIndex] == null ||
+            row[latIndex].isEmpty() ||
+            row[lonIndex] == null ||
+            row[lonIndex].isEmpty())) {
+        entityRecordsWithGeoVar++;
+      
+        double latitude = Double.parseDouble(row[latIndex]);
+        double longitude = Double.parseDouble(row[lonIndex]);
+        if (viewport.containsCoordinates(latitude, longitude)) {
+          aggregator.putIfAbsent(row[geoVarIndex], new GeoVarData());
+          aggregator.get(row[geoVarIndex]).addRow(latitude, longitude);
+        }
+        nextLine = reader.readLine();
+      }  
+    }
+
+    // begin output object and single property containing array of map elements
+    out.write("{\"mapElements\":[".getBytes(StandardCharsets.UTF_8));
+    boolean first = true;
+    for (String key : aggregator.keySet()) {
+      // write commas between elements
+      if (first) first = false; else out.write(',');
+      GeoVarData data = aggregator.get(key);
+      GeographicPoint avgLatLon = data.latLonAvg.getCurrentAverage();
+      out.write(new JSONObject()
+        .put("geoAggregateValue", key)
+        .put("entityCount", data.count)
+        .put("avgLat", avgLatLon.getLatitude())
+        .put("avgLon", avgLatLon.getLongitude())
+        .put("minLat", data.minLat)
+        .put("minLon", data.minLon)
+        .put("maxLat", data.maxLat)
+        .put("maxLon", data.maxLon)
+        .toString()
+        .getBytes(StandardCharsets.UTF_8)
+      );
+    }
+    // add config and close
```

Comment:
So is the plan to use pure Java for the overlay summary (categorical "barchart/donut" or continuous histogram)?

If so, I'd suggest sticking with R for now, while we build the client.

Reply:
the current plan is to have the quantiles/ bin ranges/ filter-aware-var-metadata endpoint do its work in R, and this one in Java. The map plugin that belongs to the pass app will keep doing its work in R. i assumed youd keep building the client from that endpoint, and switch to this one when i finally have it working. i know that means an api change, but this wont be ready for a while yet probably. theres a good amount of work to do in R as well.